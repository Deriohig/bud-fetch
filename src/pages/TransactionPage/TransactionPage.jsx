import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import fetchTransactions from '../../api/fetchTransactions.js';
import Skeleton from '../../components/SkeletonState/index';
import TransactionCard from '../../components/TransactionCard/index';
import Notifcation from '../../components/Notification/index';

const Title = styled.h3`
    font-weight: 500;
    font-size: 18px;
    color: #161616;
    font-family: Helvetica, Sans-Serif;
    margin-bottom: 24px;
`;

const PageWrapper = styled.section`
    margin: 0 auto;
    max-width: 65%;
    background-color: #fff;
    padding: 16px;
`;

const CardWrapper = styled.div`
    display: flex;
    flex-wrap: wrap;
`;

const TransactionPage = () => {
    const [loading, setLoading] = useState(true);
    const [showNotification, setShowNotification] = useState(false);
    const [notification, setNotification] = useState({});
    const [transactionsList, setTransactionsList] = useState([]);

    useEffect(() => {
        getTransactions();
    }, []);

    const filterResults = (transations) => {
        const filteredTransactions  = transations.filter((transation)=> transation.amount.value < 0);
        const sortedTransactions = filteredTransactions.sort((a, b) => b.amount.value - a.amount.value );
        if (filteredTransactions.length > 10) {
            const truncatedTransaction = sortedTransactions.splice(0, 10);
            return truncatedTransaction;
        } else return sortedTransactions;
    };

    const getTransactions = async () => {
        try {
            const response = await fetchTransactions();
            const data = await response.json();
            const filteredData = data.transactions.length ? filterResults(data.transactions) : [];
            setTransactionsList(filteredData);
            setLoading(false);
        } catch (error) {
            setLoading(false);
            setNotification({ title: error.title ? error.title : 'An error occured', kind: 'error', message: error.message });
            setShowNotification(error.message);
        }
    };

    const renderTransactionCards = () => {
        return transactionsList.map((transaction, index) => (
            <TransactionCard
                key={index}
                date={transaction.date}
                currency={transaction.amount.currency_iso}
                value={transaction.amount.value}
                description={transaction.description}
                category={transaction.category_title}
            />
        ));
    };

    return (
        <PageWrapper data-testid="TransactionPage">
            <Title>Smallest transactions</Title>
            {loading && (
                <>
                    <Skeleton width={'30%'} />
                    <Skeleton width={'70%'} />
                </>
            )}
            {!loading && (
                <CardWrapper data-testid="CardWrapper"> {renderTransactionCards()} </CardWrapper>
            )}
            {showNotification && (
                <Notifcation
                    title={notification.title}
                    handleClick={() => setShowNotification(false)}
                    message={notification.message}
                    kind={notification.kind}
                />
            )}
        </PageWrapper>
    );
};

export default TransactionPage;
