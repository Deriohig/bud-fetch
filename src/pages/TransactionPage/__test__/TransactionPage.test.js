import React from 'react';
import TransactionPage from '../TransactionPage';
import { render, cleanup, waitFor } from '@testing-library/react';
import 'regenerator-runtime/runtime';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import MockData from './__mocks__/MockData';
import EmptyData from './__mocks__/EmptyData';

afterEach(cleanup);

const getData = (data) => {
    const mockJsonPromise = Promise.resolve(data);
    const mockFetchPromise = Promise.resolve({
        json: () => mockJsonPromise,
    });
    global.fetch = jest.fn().mockImplementationOnce(() => mockFetchPromise);
}

it('renders data into cards after skeleton loader', async () => {
    getData(MockData);
    const { getByTestId, getAllByTestId } = render(<TransactionPage />);
    expect(getAllByTestId('SkeletonState')).toHaveLength(2);
    const CardWrapper = await waitFor(() => getByTestId('CardWrapper'));
    expect(CardWrapper.children).toHaveLength(10);
});

it('renders without data', async () => {
    getData(EmptyData);
    const { getByTestId, getAllByTestId } = render(<TransactionPage />);
    expect(getAllByTestId('SkeletonState')).toHaveLength(2);
    const CardWrapper = await waitFor(() => getByTestId('CardWrapper'));
    expect(CardWrapper.children).toHaveLength(0);
});

it('renders error with failed API call', async () => {
    const { getByTestId, getAllByTestId } = render(<TransactionPage />);
    expect(getAllByTestId('SkeletonState')).toHaveLength(2);
    const Notification = await waitFor(() => getByTestId('Notification'));
    expect(Notification).toBeVisible();
});

it('matches snapshot', () => {
    getData(MockData);
    const snapshot = renderer.create(<TransactionPage />).toJSON();
    expect(snapshot).toMatchSnapshot();
});
