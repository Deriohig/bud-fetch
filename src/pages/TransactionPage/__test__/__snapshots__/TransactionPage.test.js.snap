// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section
  className="sc-jrAGrp kNgSmL"
  data-testid="TransactionPage"
>
  <h3
    className="sc-pFZIQ iACuQP"
  >
    Smallest transactions
  </h3>
  <p
    className="sc-bdfBwQ knDrYM"
    data-testid="SkeletonState"
    width="30%"
  />
  <p
    className="sc-bdfBwQ cgmrII"
    data-testid="SkeletonState"
    width="70%"
  />
</section>
`;
